internal/pkg/config/config.go

```diff
@@ -155,19 +142,76 @@ func (c *Config) access() *ucfg.Config {
 
 // Merge merges two configuration together.
 func (c *Config) Merge(from interface{}, opts ...interface{}) error {
-	ucfgOpts, err := getUcfgOptions(opts...)
+	ucfgOpts, _, err := getOptions(opts...)
 	if err != nil {
 		return err
 	}
 	return c.access().Merge(from, ucfgOpts...)
 }
 
 // ToMapStr takes the config and transform it into a map[string]interface{}
-func (c *Config) ToMapStr() (map[string]interface{}, error) {
+func (c *Config) ToMapStr(opts ...interface{}) (map[string]interface{}, error) {
+	if len(opts) == 0 {
+		opts = DefaultOptions
+	}
+	ucfgOpts, local, err := getOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// remove and unpack each skip keys into its own map with no resolve
+	// so that variables are not substituted
+	skippedKeys := map[string]interface{}{}
+	skippedKeysOrig := map[string]*ucfg.Config{}
+	for _, skip := range local.skipKeys {
+		if c.access().HasField(skip) {
+			subCfg, err := c.access().Child(skip, -1)
+			if err != nil {
+				return nil, err
+			}
+			var subUnpacked interface{}
+			if subCfg.IsDict() {
+				var subDict map[string]interface{}
+				err = subCfg.Unpack(&subDict, ucfg.ResolveNOOP)
+				if err != nil {
+					return nil, err
+				}
+				subUnpacked = subDict
+			} else if subCfg.IsArray() {
+				var subArr []interface{}
+				err = subCfg.Unpack(&subArr, ucfg.ResolveNOOP)
+				if err != nil {
+					return nil, err
+				}
+				subUnpacked = subArr
+			} else {
+				return nil, fmt.Errorf("unsupported type for SkipKeys option %s", skip)
+			}
+			_, err = c.access().Remove(skip, -1)
+			if err != nil {
+				return nil, err
+			}
+			skippedKeys[skip] = subUnpacked
+			skippedKeysOrig[skip] = subCfg
+		}
+	}
+
+	// perform unpack with the skip keys removed
 	var m map[string]interface{}
-	if err := c.Unpack(&m); err != nil {
+	if err := c.access().Unpack(&m, ucfgOpts...); err != nil {
 		return nil, err
 	}
+
+	// add the skipped keys into the map and back into the config
+	for k, v := range skippedKeys {
+		m[k] = v
+	}
+	if len(skippedKeysOrig) > 0 {
+		err := c.access().Merge(skippedKeysOrig, ucfg.ResolveNOOP)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
I am confused by this block: after adding in the resulting map the skipped keys without expanding their values why do we merge it back into config with the `${<name>}` ?
Moreover it seems strange for a method that creates a `map[string]interface{}` out of a Config object to have side effects, is there is a specific reason for that ?

Reply:
That is because the code above removed the entire field from the original `*Config` (aka. it mutated), so it must add them back in or the configuration after this point would be missing those fields.

Personally I hate this code, this really should be added as a core feature in `go-ucfg` which would result in this not having all this hacking stuff in place.

Without this change it is impossible for the `ToMapString()` to work, because once it hits a variable it will unpack it and try to resolve it, breaking it.